### PR TITLE
[CN-412] Add E2E tests for external backup type

### PR DESCRIPTION
**Pull request by dzeromski-hazelcast**

### CODE REVIEW

File: test/e2e/config/hazelcast/config.go [hunk 1]
```diff
@@ -188,6 +188,54 @@ var (
...
+				Version:          naming.HazelcastVersion,
+				LicenseKeySecret: licenseKey(ee),
+				Persistence: &hazelcastv1alpha1.HazelcastPersistenceConfiguration{
+					BaseDir:                   "/data/hot-restart",
+					ClusterDataRecoveryPolicy: hazelcastv1alpha1.FullRecovery,
+					Pvc: hazelcastv1alpha1.PersistencePvcConfiguration{
+						AccessModes:    []corev1.PersistentVolumeAccessMode{corev1.ReadWriteOnce},
+						RequestStorage: &[]resource.Quantity{resource.MustParse("8Gi")}[0],
+					},
+				},
```
Comment (cagric0) [inline test/e2e/config/hazelcast/config.go:210]: Don't we need [_BackupType_](https://github.com/hazelcast/hazelcast-platform-operator/blob/5425ff1760a413593cd72b24813ac95be865f632/api/v1alpha1/hazelcast_types.go#L159) field inside the Persistence

File: test/e2e/config/hazelcast/config.go [hunk 2]
```diff
@@ -188,6 +188,54 @@ var (
...
```
Comment (cagric0) [inline test/e2e/config/hazelcast/config.go:234]: We need RestoreConfiguration under the Persistence.

Comment (cagric0) [inline test/e2e/config/hazelcast/config.go:234]: We need RestoreConfiguration under the Persistence.

File: test/e2e/hazelcast_persistence_test.go [hunk 3]
```diff
@@ -244,4 +244,27 @@ var _ = Describe("Hazelcast CR with Persistence feature enabled", Label("hz_pers
...
+		}
+
+		By("Create cluster with external backup enabled")
+		hazelcast := hazelcastconfig.ExternalBackup(hzLookupKey, false, labels)
+		CreateHazelcastCR(hazelcast)
+		evaluateReadyMembers(hzLookupKey, 3)
+
+		By("Trigger backup")
+		hotBackup := hazelcastconfig.HotBackup(hbLookupKey, hazelcast.Name, labels)
+		Expect(k8sClient.Create(context.Background(), hotBackup)).Should(Succeed())
```
Comment (cagric0) [inline test/e2e/hazelcast_persistence_test.go:260]: Before triggering hotBackup you will seed data into HZ right?

Comment (cagric0) [inline test/e2e/hazelcast_persistence_test.go:260]: Before triggering hotBackup you will seed data into HZ right?

**Review by cagric0 — COMMENTED** [review]
I add some comments but I think you already know them, you're just sending small commits to make the PR easy to review. Thanks for that!

File: test/e2e/config/hazelcast/config.go (hunk 1 above)
Comment (dzeromski-hazelcast) [reply test/e2e/config/hazelcast/config.go]: Yup.

File: test/e2e/hazelcast_persistence_test.go (hunk 3 above)
Comment (dzeromski-hazelcast) [reply test/e2e/hazelcast_persistence_test.go]: Not in this test

File: test/e2e/config/hazelcast/config.go [hunk 4]
```diff
@@ -188,6 +188,82 @@ var (
...
+				Labels:    labels,
+			},
+			Spec: hazelcastv1alpha1.HazelcastSpec{
+				ClusterSize:      &[]int32{1}[0],
+				Repository:       repo(ee),
+				Version:          naming.HazelcastVersion,
+				LicenseKeySecret: licenseKey(ee),
+				Agent: &hazelcastv1alpha1.AgentConfiguration{
+					Repository: "dzeromskhazelcast/platform-operator-agent",
+					Version:    "latest",
```
Comment (cagric0) [inline test/e2e/config/hazelcast/config.go:205]: Why aren't we using `hazelcast/platform-operator-agent` and its release version `0.1.0`

File: test/e2e/config/hazelcast/config.go [hunk 5]
```diff
@@ -188,6 +188,82 @@ var (
...
+				},
+				Persistence: &hazelcastv1alpha1.HazelcastPersistenceConfiguration{
+					BaseDir:                   "/data/hot-restart",
+					ClusterDataRecoveryPolicy: hazelcastv1alpha1.FullRecovery,
+					Pvc: hazelcastv1alpha1.PersistencePvcConfiguration{
+						AccessModes:    []corev1.PersistentVolumeAccessMode{corev1.ReadWriteOnce},
+						RequestStorage: &[]resource.Quantity{resource.MustParse("8Gi")}[0],
+					},
+					Restore: &hazelcastv1alpha1.RestoreConfiguration{
+						BucketURI: "s3://hazelcast-cn-306-restore-tests?prefix=" + prefix,
```
Comment (cagric0) [inline test/e2e/config/hazelcast/config.go:244]: Are we gonna use this hardcoded bucket?

**Review by cagric0 — APPROVED** [review]
I've a few questions other than that LGTM!

File: test/e2e/hazelcast_persistence_test.go [hunk 6]
```diff
@@ -244,4 +245,55 @@ var _ = Describe("Hazelcast CR with Persistence feature enabled", Label("hz_pers
 		Entry("with HostPath configuration single node", Label("slow"), "/tmp/hazelcast/singleNode", "dummyNodeName"),
 		Entry("with HostPath configuration multiple nodes", Label("slow"), "/tmp/hazelcast/multiNode"),
 	)
+
+	FIt("should successfully restore from external backup", Label("slow"), func() {
```
Comment (SeriyBg) [inline test/e2e/hazelcast_persistence_test.go:249]:
```suggestion
	It("should successfully restore from external backup", Label("slow"), func() {
```

File: test/e2e/config/hazelcast/config.go (hunk 4 above)
Comment (dzeromski-hazelcast) [reply test/e2e/config/hazelcast/config.go:205]: I will change that and update default agent version.

File: test/e2e/hazelcast_persistence_test.go (hunk 6 above)
Comment (dzeromski-hazelcast) [reply test/e2e/hazelcast_persistence_test.go:249]: I'm changing workflows and wanted to reduce test time. This is temporary.

File: test/e2e/hazelcast_persistence_test.go [hunk 7]
```diff
@@ -244,4 +245,55 @@ var _ = Describe("Hazelcast CR with Persistence feature enabled", Label("hz_pers
...
+
+		logs := InitLogs(t, hzLookupKey)
+		defer logs.Close()
+		scanner := bufio.NewScanner(logs)
+		test.EventuallyInLogs(scanner, 1*Second, logInterval).Should(ContainSubstring("Found existing hot-restart directory"))
+		test.EventuallyInLogs(scanner, 1*Second, logInterval).Should(ContainSubstring("Local Hot Restart procedure completed with success."))
+	})
 })
+
+func unixMilli(msec int64) Time {
```
[the remark this answers is not in the document]
Comment (dzeromski-hazelcast) [reply test/e2e/hazelcast_persistence_test.go:297]: Backup sequence numbers are in milliseconds so this is just a helper to convert it to proper timestamp.

File: test/e2e/config/hazelcast/config.go (hunk 5 above)
Comment (dzeromski-hazelcast) [reply test/e2e/config/hazelcast/config.go:244]: I will convert test to DescribeTable and use bucket as parameter.

File: test/e2e/hazelcast_persistence_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (SeriyBg) [reply test/e2e/hazelcast_persistence_test.go:297]: Oh, I see, it's just the copy of [this](https://cs.opensource.google/go/go/+/refs/tags/go1.18.3:src/time/time.go;l=1349). We need to update our Golang version :)

[the remark this answers is not in the document]
Comment (dzeromski-hazelcast) [reply test/e2e/hazelcast_persistence_test.go:299]: Can I do it in separate PR? I dont want to break something and revert this :)

[the remark this answers is not in the document]
Comment (SeriyBg) [reply test/e2e/hazelcast_persistence_test.go:299]: Oh, sorry, I've meant that we need to upgrade the Golang version in general. Definitely not in the scope of this PR 😄

**Review by SeriyBg — APPROVED** [review]

**Pull request merged**